src/inmanta/agent/handler.py

```diff
@@ -961,6 +977,109 @@ def execute(self, ctx: HandlerContext, resource: TPurgeableResource, dry_run: bo
         super().execute(ctx, resource, dry_run)
 
 
+@stable_api
+class DiscoveryHandler(HandlerABC, Generic[DRT, URT]):
+    # The DiscoveryHandler is generic in both the handler's Discovery Resource type (DRT)
+    # and the Unmanaged Resource type (URT) it reports to the server. The second has to be serializable.
+
+    # This class deploys instances of DRT and reports URT to the server.
+
+    def __init__(self, agent: "inmanta.agent.agent.AgentInstance") -> None:
+        self._agent = agent
+        self._ioloop = agent._io_loop
+
+    @abstractmethod
+    def discover_resources(self, ctx: HandlerContext, discovery_resource: DRT) -> abc.Mapping[ResourceIdStr, URT]:
+        raise NotImplementedError
+
+    def report_discovered_resources(self, ctx: HandlerContext, resource: DRT) -> None:
+        """ """
+
+        def _call_discovered_resource_create_batch() -> typing.Awaitable[Result]:
+            return self.get_client().discovered_resource_create_batch(
+                tid=self._agent.environment, discovered_resources=discovered_resources
+            )
+
+        try:
+            self.pre(ctx, resource)
+            # report to the server
+            discovered_resources: List[DiscoveredResource] = [
+                DiscoveredResource(discovered_resource_id=resource_id, values=values)
+                for resource_id, values in self.discover_resources(ctx, resource).items()
+            ]
+            result = self.run_sync(_call_discovered_resource_create_batch)
+
+            if result.code != 200:
+                error_msg_from_server = f": {result.result['message']}" if "message" in result.result else ""
+                raise Exception(f"Failed to report discovered resources to the server{error_msg_from_server}")
+
+        except Exception as e:
+            ctx.set_status(const.ResourceState.failed)
+            ctx.exception(
+                (
+                    "An error occurred during resource discovery of type %(urt)s, "
+                    "triggered by %(resource_id)s (exception: %(exception)s"
+                ),
+                urt=str(URT),
+                resource_id=resource.id,
+                exception=f"{e.__class__.__name__}('{e}')",
+            )
+        finally:
+            try:
+                self.post(ctx, resource)
+            except Exception as e:
+                ctx.exception(
+                    (
+                        "An error occurred after resource discovery of type %(urt)s, "
+                        "triggered by %(resource_id)s (exception: %(exception)s"
+                    ),
+                    urt=str(URT),
```

Comment:
`URT` is just a `TypeVar`, not the bound type. It is also just an intermediate representation type so I'm not sure if it is even relevant to the end user. Same for the other message